helm kubernetes example based on kind cluster and nginx ingress

EvgeniGordeev opened the pull request:
A kubernetes example based on kind, nginx-ingress, and helm (optional).

JoelSpeed commented on contrib/helm/Makefile:
```diff
@@ -0,0 +1,44 @@
+
+.PHONY: setup-ingress
+setup-ingress:
+	kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/static/provider/kind/deploy.yaml
```
Since this links to master, it's liable to change, should we be pinning it instead?

EvgeniGordeev replied:
I followed their example https://kind.sigs.k8s.io/docs/user/ingress/#ingress-nginx. It makes sense not to pin since we don't pin the version of kind. Installing the latest kind is expected to work with their documentation.

JoelSpeed commented on contrib/helm/Makefile:
```diff
@@ -0,0 +1,44 @@
+	kubectl --namespace ingress-nginx wait --for=condition=available --timeout=30s deployment/ingress-nginx-controller
+	kubectl --namespace ingress-nginx rollout status --timeout 90s deployment/ingress-nginx-controller
+
+# default Pod CIDR is 10.244.0.0/16 https://github.com/kubernetes-sigs/kind/blob/master/pkg/apis/config/v1alpha4/default.go
```
Again, maybe permalink rather than pinning to master as this will be liable to change?

JoelSpeed commented on contrib/helm/Makefile:
```diff
@@ -0,0 +1,44 @@
+	kubectl -n kube-system rollout restart deployment/coredns
+	kubectl -n kube-system rollout status --timeout 60s deployment/coredns
+
+.PHONY: undeploy
```
Typo I believe
```suggestion
.PHONY: delete-cluster
```

JoelSpeed commented on contrib/helm/Makefile:
```diff
@@ -0,0 +1,44 @@
+helm-init:
+	helm dep update
+
+.PHONY: undeploy
```
Typo
```suggestion
.PHONY: deploy
```

JoelSpeed commented on contrib/helm/hello-world/values.yaml:
```diff
@@ -0,0 +1,79 @@
+  repository: tutum/hello-world
+  pullPolicy: IfNotPresent
+  # Overrides the image tag whose default is the chart version.
+  tag: latest
```
Should this be pinned?

EvgeniGordeev replied:
created a separate version of hello-world with pinned tag.

EvgeniGordeev commented:
> Perhaps the contrib/helm folder should actually be contrib/local-environment/kubernetes instead
Yes, that what I was working yesterday on - actually making helm optional and leave it only to kubectl and kind as the required deps and then renaming the directory to a generic `kubernetes`. I'll check out all your comments before making the new commit here. 
Another idea I wanted to add was to make a hello-world more useful by displaying actual request headers and cookies to give an overview what to expect.

JoelSpeed commented on contrib/helm/Makefile:
```diff
@@ -0,0 +1,44 @@
+	kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/static/provider/kind/deploy.yaml
+	sleep 5 # because of https://github.com/kubernetes/kubernetes/issues/83242
+	kubectl --namespace ingress-nginx wait --for=condition=available --timeout=30s deployment/ingress-nginx-controller
+	kubectl --namespace ingress-nginx rollout status --timeout 90s deployment/ingress-nginx-controller
```
When I tried this earlier, 90s was too short and it errored out for me. Is there any harm upping this to say 5m which would be inline with the help deployments?

EvgeniGordeev replied:
sounds good, will do. One of the thing I prefer helm is that it has a built-in readiness check whereas with kubectl I had to insert those checks.

NickMeves replied:
What's the rationale for applying the raw ingress resources via kubectl instead of using the `stable/nginx-ingress` helm chart since we are using helm in most the other places?

EvgeniGordeev replied:
I tried it first :) but helm chart as is didn't work for kind. More athttps://github.com/kubernetes/ingress-nginx/issues/5569. Using the nginx-ingress chart will require a ton of workarounds applied on top of the chart.

NickMeves replied:
That likely explains the tons of special values I have around the `stable/nginx-ingress` in our kind scripts for my work environments 😄

EvgeniGordeev replied:
In the end values.yaml are expected to contain charts relevant to the example and not to provide a complete guide how to customize nginx-ingress :) .

NickMeves commented on contrib/helm/kind-cluster.yaml:
```diff
@@ -0,0 +1,14 @@
+          kubeletExtraArgs:
+            node-labels: "ingress-ready=true"
+    extraPortMappings:
+      - containerPort: 80
```
Should we open up 443 to the kind cluster as well in case users want to test the ingress with TLS in the cluster?

EvgeniGordeev replied:
I wanted to as least invasive as possible on the host machine of someone trying this out. I can add 443 as well if you think it makes sense. @JoelSpeed thoughts?

EvgeniGordeev commented:
* moved to `contrib/local-environment/kubernetes`
* added `kubernetes-up`/`-down` targets to `contrib/local-environment/Makefile`
* moved out httpbin and hello-world charts outside of this project to save on boilerplate code

JoelSpeed commented on contrib/local-environment/Makefile:
```diff
@@ -21,3 +21,12 @@ keycloak-up:
+
+.PHONY: kubernetes-up
+kubernetes-up:
+	cd kubernetes && make create-cluster
+	cd kubernetes && make deploy
```
Does `make -C kubernetes` work instead? That's what we have in other placed where we are calling makefiles in different directories

JoelSpeed commented on contrib/local-environment/kubernetes/Makefile:
```diff
@@ -0,0 +1,66 @@
+# creates K8S manifest from helm chart
+.PHONY: helm-create-manifest
+helm-create-manifest: helm-init
+	helm template --namespace default oauth2-proxy-example . > oauth2-proxy-example-full.yaml
```
Might be good to add a header to the file that we are generating here to explain it was autogenerated?
```suggestion
	echo "# WARNING: This file is auto-generated by `make helm-create-manifest`! DO NOT EDIT MANUALLY!" > oauth2-proxy-example-full.yaml
	helm template --namespace default oauth2-proxy-example . >> oauth2-proxy-example-full.yaml
```

EvgeniGordeev commented:
Suggestions pushed.

EvgeniGordeev commented:
@JoelSpeed I ran into an issue today when I updated `cookie_domain=".localtest.me"` to `cookie_domains=[".localtest.me"]`. 

With a single domain `cookie_domain` oauth2-proxy in the callback sets the cookie `_oauth2_proxy` on domain `localtest.me` whereas with `cookie_domains` the domain is `oauth2-proxy.localtest.me` which doesn't work any more. Is it the expected behavior?

JoelSpeed commented:
@EvgeniGordeev Sounds like you worked out what was wrong? We made a bunch of changes this release, one of those being to rename `cookie_domain` to `cookie_domains`, which will work in the config from the next release v6.0.0 (imminently coming).

EvgeniGordeev commented:
@JoelSpeed yeah, I worked it out. I am going to include the necessary updates once the new version is out there while adding a keycloak example.
